counsel.el

```diff
@@ -6388,7 +6394,47 @@ sub-directories that builds may be invoked in."
   (let ((srcdir (counsel--compile-root)))
     (when (directory-files (or blddir srcdir) nil
                            counsel-compile-make-pattern t)
-      (counsel--compile-get-make-targets srcdir blddir))))
+      (counsel--compile-get-make-targets
+       #'counsel-compile--probe-make-targets srcdir blddir))))
+
+
+(defun counsel-compile--probe-make-help (dir)
+  "Return a list of Make targets based on help for `DIR'.
+
+It is quite common for a 'make help' invocation to return a human
+readable list of targets.  Often common targets are marked with a
+leading asterisk.  The exact search pattern is controlled by
+`counsel-compile-help-pattern'."
+  (let* ((default-directory dir)
+         (c-env (split-string
+                 (substring-no-properties
+                  (mapconcat #'identity counsel-compile-env " "))))
+         (p-res (apply #'call-process "make" nil nil nil "-q" "help" c-env)))
+    ;; only proceed if the help target exists
+    (when (and (numberp p-res) (= p-res 1))
+      (with-temp-buffer
+        (let (primary-targets targets)
+          (setq p-res
+                (apply #'call-process "make" nil t nil "help" c-env))
+          (when (and (numberp p-res) (= p-res 0))
+            (goto-char (point-min))
+            (while (re-search-forward counsel-compile-help-pattern nil t)
+              (push (match-string-no-properties 2)
+                    (if (string= "*" (match-string-no-properties 1))
+                        primary-targets targets)))
+            (nconc (sort primary-targets #'string-lessp)
+                   (sort targets #'string-lessp))))))))
```

Comment:
Combined with my previous comment about `match-beginning`, can this be simplified as follows?

```el
(let ((default-directory dir)
      primary-targets targets)
  ;; Only proceed if the help target exists.
  (when (eql 1 (apply #'call-process "make" nil nil nil "-q" "help"
                      counsel-compile-env))
    (with-temp-buffer
      (when (eql 0 (apply #'call-process "make" nil t nil "help"
                          counsel-compile-env))
        (goto-char (point-min))
        (while (re-search-forward counsel-compile-help-pattern nil t)
          (push (match-string 2)
                (if (match-beginning 1) primary-targets targets)))
        (nconc (sort primary-targets #'string-lessp)
               (sort targets #'string-lessp))))))
```

Reply:
@basil-conto: do properties just disappear when strings are sent outside of Emacs?

Reply:
> do properties just disappear when strings are sent outside of Emacs?

Text properties are a mostly transparent Emacs-specific data representation, so yes.

Even within Emacs, two strings with the same contents but different properties are still `equal`.

In general you shouldn't need to worry about application-specific properties like the ones Counsel uses, as they have no visible effects on other parts of Emacs (unless there's a clash of property names).

Reply:
That code tested beautifully ;-)